### PR TITLE
[RFC FS-1068] open static classes 

### PR DESCRIPTION
**Pull request by dsyme**
Continuation of https://github.com/dotnet/fsharp/pull/6309

Proof of concept for [RFC FS-1068 Open static classes](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1068-open-static-classes.md). 

Details are on the RFC, but for example:

```fsharp
> open System.Math;;
> Min(1.0, 2.0);;
val it : float = 1.0

> Min(2.0, 1.0);;
val it : float = 1.0

> open System.Math;;
> Min(2.0, 1.0);;
val it : float = 1.0
```

DONE:
* [x] Resolve how static classes are defined in F#
* [x] Test opening classes with `AutoOpen` attribute
* [x] Respect and test for `RequireQualifiedNameAttribute(true)` on a static class, that prevents it being opened, or at least gives a warning, as for F# modules.
* [x] Check completions on "open System.Ma$"
* [x] Resolve [`open` v `open static`](https://github.com/fsharp/fslang-design/issues/352#issuecomment-499146012)

### CODE REVIEW

**Comment by cartermp** [comment]
@dsyme Do the two TODOs make this a WIP?

**Comment by dsyme** [comment]
@cartermp Yes, though it should be very quick to resolve these and close this out.

**Comment by cartermp** [comment]
Tagging @TIHan for review

File: src/buildfromsource/FSharp.Compiler.Interactive.Settings/InternalsVisibleTo.fs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
```
Comment (cartermp) [inline src/buildfromsource/FSharp.Compiler.Interactive.Settings/InternalsVisibleTo.fs]: Remove this file

File: vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs [hunk 2]
```diff
@@ -3639,8 +3639,8 @@ let x = query { for bbbb in abbbbc(*D0*) do
         AssertAutoCompleteContains 
             [ "open System." ]
             "." // marker
-            [ "Collections" ] // should contain (namespace)
-            [ "Console" ] // should not contain (type)
+            [ "Collections"; "Console" ] // should contain (namespace, static type)
```
Comment (cartermp) [inline vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs]: ugh these legacy tests make me sad

File: src/fsharp/NameResolution.fs [hunk 3]
```diff
@@ -417,6 +417,12 @@ type NameResolutionEnv =
 // Helpers to do with extension members
 //-------------------------------------------------------------------------
 
+/// Indicates if we only need one result or all possible results from a resolution.
+[<RequireQualifiedAccess>]
+type ResultCollectionSettings =
+| AllResults
```
Comment (cartermp) [inline src/fsharp/NameResolution.fs]: please indent this by four spaces

File: src/fsharp/NameResolution.fs [hunk 4]
```diff
@@ -488,6 +494,125 @@ let private GetCSharpStyleIndexedExtensionMembersForTyconRef (amap: Import.Impor
...
+    let amap = infoReader.amap
+    let pinfos = GetIntrinsicPropInfoSetsOfType infoReader optFilter ad AllowMultiIntfInstantiations.Yes findFlag m ty
+    let pinfos = pinfos |> ExcludeHiddenOfPropInfos g amap m
+    pinfos
+
+/// Select from a list of extension properties
+let SelectPropInfosFromExtMembers (infoReader:InfoReader) ad optFilter declaringTy m extMemInfos =
+    let g = infoReader.g
+    let amap = infoReader.amap
+    // NOTE: multiple "open"'s push multiple duplicate values into eIndexedExtensionMembers, hence setify.
```
Comment (cartermp) [inline src/fsharp/NameResolution.fs]: `List.setify` isn't used here, is this comment outdated?

File: vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs (hunk 2 above)
Comment (cartermp) [reply vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs]: (this comment is not actionable 🙂)

**Comment by dsyme** [comment]
@cartermp @TIHan We have one remaining design issue for this feature, https://github.com/fsharp/fslang-design/issues/352#issuecomment-499146012, basically `open` v. `open static`

**Comment by dsyme** [comment]
> @cartermp @TIHan We have one remaining design issue for this feature, fsharp/fslang-design#352 (comment), basically open v. open static

I have resolved this (https://github.com/fsharp/fslang-design/issues/352#issuecomment-508140249), and this PR is now ready for review 

@KevinRansom @TIHan we need to add language version flag protection for this feature.  What's our status for doing this?  I'm happy to do it but it needs to be done ASAP

Also should this go into `feature/fsharp47` or `master`?  Again happy to retarget if necessary

**Comment by dsyme** [comment]
@cartermp please re-review when you get the chance

**Comment by dsyme** [comment]
@KevinRansom @TIHan Please advise on adding the `/langversion` checks for this PR. This needs to be done asap, thanks.  I can do it, just let me know.

**Comment by cartermp** [comment]
@dsyme if you have the time, feel free to guard this with the langversion switch without awaiting @KevinRansom. We're still getting our ducks in row for 16.3 :)

**Comment by dsyme** [comment]
@KevinRansom @cartermp I rebased to `release/fsharp47` instead of `master`

File: src/fsharp/service/FSharpCheckerResults.fs [hunk 5]
```diff
@@ -850,7 +850,12 @@ type internal TypeCheckInfo
             | Some(CompletionContext.OpenDeclaration) ->
                 GetDeclaredItems (parseResultsOpt, lineStr, origLongIdentOpt, colAtEndOfNamesAndResidue, residueOpt, lastDotPos, line, loc, filterCtors, resolveOverloads, hasTextChangedSinceLastTypecheck, false, getAllSymbols)
                 |> Option.map (fun (items, denv, m) ->
-                    items |> List.filter (fun x -> match x.Item with Item.ModuleOrNamespaces _ -> true | _ -> false), denv, m)
+                    items 
+                    |> List.filter (fun x ->
+                        match x.Item with
+                        | Item.ModuleOrNamespaces _ -> true
+                        | Item.Types (_, tcrefs) when tcrefs |> List.exists (fun ty -> isAppTy g ty && isStaticClass g (tcrefOfAppTy g ty)) -> true
```
Comment (cartermp) [inline src/fsharp/service/FSharpCheckerResults.fs]: Not sure if this is actionable, so feel free to ignore...

This is the correct change for the feature given how this routine is written, but I wonder if we should be thinking about refactoring things like this throughout the service/FSharpCheckerResults layer. Embedding details about validity like this makes it challenging (for me, at least), to make sense of this overall routine. The parameter lists pattern even more so.

File: src/fsharp/NameResolution.fs [hunk 6]
```diff
@@ -1943,126 +2141,6 @@ let ResolveObjectConstructor (ncenv: NameResolver) edenv m ad ty =
 // Bind the "." notation (member lookup or lookup in a type)
 //-------------------------------------------------------------------------
 
-/// Query the declared properties of a type (including inherited properties)
-let IntrinsicPropInfosOfTypeInScope (infoReader:InfoReader) optFilter ad findFlag m ty =
```
Comment (cartermp) [inline src/fsharp/NameResolution.fs]: Were these functions just moved up, or modified?

File: src/fsharp/NameResolution.fs [hunk 7]
```diff
@@ -630,6 +755,30 @@ let AddUnionCases2 bulkAddMode (eUnqualifiedItems: LayeredMap<_, _>) (ucrefs: Un
             let item = Item.UnionCase(GeneralizeUnionCaseRef ucref, false)
             acc.Add (ucref.CaseName, item))
 
+let AddStaticContentOfTyconRefToNameEnv (g:TcGlobals) (amap: Import.ImportMap) m (nenv: NameResolutionEnv)  (tcref:TyconRef) =
```
Comment (cartermp) [inline src/fsharp/NameResolution.fs]: Is this the primary thing that was added?

Comment (dsyme) [reply src/fsharp/NameResolution.fs]: Yes

File: src/fsharp/service/FSharpCheckerResults.fs (hunk 5 above)
Comment (dsyme) [reply src/fsharp/service/FSharpCheckerResults.fs]: Yeah, I agree.  The whole of this routine is poorly written and has grown with adhoc cases over time. One reason I split `service.fs` into two was to start to make progress on this.

**Comment by dsyme** [comment]
@KevinRansom I resolved the conflicts, you might want to double check

**Comment by dsyme** [comment]
Ugh I messed up the merge, 
```
C:\GitHub\dsyme\visualfsharp2\tests\FSharp.Core.UnitTests\FSharp.Core\Microsoft.FSharp.Core\NameOfTests.fs(19,22): erro
r FS0039: The value or constructor 'nameof' is not defined. [C:\GitHub\dsyme\visualfsharp2\tests\FSharp.Core.UnitTests\
FSharp.Core.UnitTests.fsproj]
C:\GitHub\dsyme\visualfsharp2\tests\FSharp.Core.UnitTests\FSharp.Core\Microsoft.FSharp.Core\NameOfTests.fs(21,34): erro
r FS0039: The value or constructor 'nameof' is not defined. [C:\GitHub\dsyme\visualfsharp2\tests\FSharp.Core.UnitTests\
FSharp.Core.UnitTests.fsproj]
```

**Comment by dsyme** [comment]
We have a heisen-error in the Linux Source Build:
```
.packages/microsoft.sourcelink.common/1.0.0-beta2-19351-01/build/InitializeSourceControlInformation.targets(3,81): error MSB4022: The result "" of evaluating the value "$(_MicrosoftSourceLinkCommonAssemblyFile)" of the "AssemblyFile" attribute in element <UsingTask> is not valid.
.packages/microsoft.sourcelink.common/1.0.0-beta2-19351-01/build/InitializeSourceControlInformation.targets(3,81): error MSB4022: The result "" of evaluating the value "$(_MicrosoftSourceLinkCommonAssemblyFile)" of the "AssemblyFile" attribute in element <UsingTask> is not valid.
Build failed (exit code '1').
Bash exited with code '1'.
```

**Pull request merged**